File: Related Topics/Two Pointers/141. Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // int pos = -1;
        ListNode *p1 = head;
        ListNode *p2 = head;
        // bool flag = false;
        while(p2 && p2->next){
            // pos += 1;
            p1 = p1->next;
            p2 = p2->next->next;
            if(p1 == p2){
                // flag = true;
                // break;
                return true;
            }
        }
        /*
        if(flag){
            return true;
        }else{
            return false;
        }
        */
        return false;
    }
};

File: Related Topics/Two Pointers/141. Linked List Cycle.py
# 141. Linked List Cycle
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# class Solution:
#     def hasCycle(self, head: ListNode) -> bool:
#         if not head or not head.next:
#             return None
#         # slow, fast = ListNode(0), ListNode(0)
#         slow = head
#         fast = head.next

#         while slow != fast:
#             if not fast or not fast.next:
#                 return False
#             slow = slow.next
#             fast = fast.next.next

#         return True

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False
